src/main/java/com/amazon/dlic/auth/http/saml/HTTPSamlAuthenticator.java

```diff
@@ -171,26 +170,26 @@ public String getType() {
     }
 
     @Override
-    public boolean reRequestAuthentication(RestChannel restChannel, AuthCredentials authCredentials) {
+    public BytesRestResponse reRequestAuthentication(RestRequest request, AuthCredentials credentials) {
         try {
-            RestRequest restRequest = restChannel.request();
-            Matcher matcher = PATTERN_PATH_PREFIX.matcher(restRequest.path());
+            Matcher matcher = PATTERN_PATH_PREFIX.matcher(request.path());
             final String suffix = matcher.matches() ? matcher.group(2) : null;
-            if (API_AUTHTOKEN_SUFFIX.equals(suffix) && this.authTokenProcessorHandler.handle(restRequest, restChannel)) {
-                return true;
+            if (API_AUTHTOKEN_SUFFIX.equals(suffix)) {
+                final BytesRestResponse restResponse = this.authTokenProcessorHandler.handle(request);
+                if (restResponse != null) {
```

Comment:
The logic is a little confusing to follow here, but I see why you had to write it this way. If I am reading this correctly, does `restResponse === null` indicate that the request authenticated?

This is also specific to how SAML does auth where `reRequestAuthentication` is used to validate the XML SAML response and generate a JWT?

Reply:
yes, the implementation is not straightforward for this one.
`restResponse == null` here indicates the request was not authenticated, so proceed with the challenge. It is slightly confusing to understand, but if you go through the whole request flow, it might be easier.